// @react-md/utils/lib/keyboardMovement/movementContext.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useKeyboardFocusContext = exports.KeyboardMovementContextProvider = exports.DEFAULT_RTL_KEYBOARD_MOVEMENT = exports.DEFAULT_LTR_KEYBOARD_MOVEMENT = exports.DEFAULT_KEYBOARD_MOVEMENT = void 0;
var react_1 = require("react");
/**
 * @remarks \@since 5.0.0
 * @internal
 */
var noop = function () {
    if (process.env.NODE_ENV !== "production") {
        throw new Error("KeyboardMovementProvider must be a parent component.");
    }
};
/**
 * Most custom keyboard functionality use these keys.
 *
 * @remarks \@since 5.0.0
 * @internal
 */
exports.DEFAULT_KEYBOARD_MOVEMENT = {
    incrementKeys: ["ArrowDown"],
    decrementKeys: ["ArrowUp"],
    jumpToFirstKeys: ["Home"],
    jumpToLastKeys: ["End"],
};
/**
 * @remarks \@since 5.1.2
 * @internal
 */
exports.DEFAULT_LTR_KEYBOARD_MOVEMENT = {
    incrementKeys: ["ArrowRight"],
    decrementKeys: ["ArrowLeft"],
    jumpToFirstKeys: ["Home"],
    jumpToLastKeys: ["End"],
};
/**
 * @remarks \@since 5.1.2
 * @internal
 */
exports.DEFAULT_RTL_KEYBOARD_MOVEMENT = {
    incrementKeys: ["ArrowLeft"],
    decrementKeys: ["ArrowRight"],
    jumpToFirstKeys: ["Home"],
    jumpToLastKeys: ["End"],
};
/**
 * @remarks \@since 5.0.0
 * @internal
 */
var context = (0, react_1.createContext)({
    attach: noop,
    detach: noop,
    watching: { current: [] },
    loopable: false,
    searchable: false,
    horizontal: false,
    includeDisabled: false,
    config: { current: exports.DEFAULT_KEYBOARD_MOVEMENT },
});
context.displayName = "KeyboardMovement";
/**
 * @remarks \@since 5.0.0
 * @internal
 */
exports.KeyboardMovementContextProvider = context.Provider;
/**
 * @remarks \@since 5.0.0
 * @internal
 */
function useKeyboardFocusContext() {
    return (0, react_1.useContext)(context);
}
exports.useKeyboardFocusContext = useKeyboardFocusContext;
